tests/test_crossdock.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (c) 2016 Uber Technologies, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+
+import time
+
+import mock
+import opentracing
+import pytest
+import tornado.web
+import crossdock.server.api as api
+from crossdock.server import server
+from crossdock.server.api import Request, Downstream
+from jaeger_client import Tracer, ConstSampler
+from jaeger_client.reporter import InMemoryReporter
+from tchannel import TChannel
+from tornado.httpclient import HTTPRequest
+
+
+def test_api_to_json():
+    r = Request(serverRole='s2',
+                downstream=Downstream(
+                    serviceName='serviceName',
+                    serverRole='s3',
+                    encoding='json',
+                    hostPort='localhost:123',
+                    downstream=Downstream(
+                        serviceName='serviceName',
+                        serverRole='s4',
+                        encoding='json',
+                        hostPort='localhost:123',
+                        downstream=None)))
+    api.namedtuple_to_dict(r)
+
+
+@pytest.fixture
+def app():
+    """Required by pytest-tornado's http_server fixture"""
+    return tornado.web.Application()
+
+
+# noinspection PyShadowingNames
+@pytest.yield_fixture
+def tracer():
+    tracer = Tracer(
+        service_name='test-tracer',
+        sampler=ConstSampler(True),
+        reporter=InMemoryReporter(),
+    )
+    try:
+        yield tracer
+    finally:
+        tracer.close()
+
+
+@pytest.mark.parametrize('s2_encoding,s3_encoding,sampled', [
+    ('json', 'json', True),
```

Comment:
I wonder if enumerating all of these permutations could be done programmatically.